platform/admin/core/admin-core-appservice/src/main/java/org/codice/ddf/admin/application/service/impl/ApplicationServiceBean.java

```diff
@@ -362,96 +366,84 @@ public void removeApplication(String appName) {
+
     if (services.isEmpty()) {
-      return returnValues;
+      return Collections.emptyList();
     }
 
     Set<BundleInfo> bundleInfos = getBundleInfosForApplication(applicationID);
     if (bundleInfos == null) {
-      return returnValues;
+      return Collections.emptyList();
     }
 
-    Set<String> bundleLocations = new HashSet<>();
-    for (BundleInfo each : bundleInfos) {
-      bundleLocations.add(each.getLocation());
-    }
+    Set<String> bundleLocations =
+        bundleInfos.stream().map(BundleInfo::getLocation).collect(Collectors.toSet());
 
     MetaTypeService metatypeService = getMetaTypeService();
-    Set<MetaTypeInformation> metatypeInformation = new HashSet<>();
-    if (metatypeService != null) {
-      for (Bundle each : getContext().getBundles()) {
-        for (BundleInfo bundleInfo : bundleInfos) {
-          if (bundleInfo.getLocation().equals(each.getLocation())) {
-            metatypeInformation.add(metatypeService.getMetaTypeInformation(each));
-          }
-        }
-      }
-    }
-
-    for (Map<String, Object> each : services) {
-      if (each.containsKey("configurations")) {
-        List<Map<String, Object>> configurations =
-            (List<Map<String, Object>>) each.get("configurations");
-        for (Map<String, Object> item : configurations) {
-          if (item.containsKey("bundle_location")) {
-            String bundleLocation = (String) item.get("bundle_location");
-            if (bundleLocations.contains(bundleLocation)) {
-              returnValues.add(each);
-              break;
-            }
-          }
-        }
-      } else {
-        if (checkForMetaTypesForService(metatypeInformation, each)) {
-          returnValues.add(each);
-        }
-      }
-    }
+    Set<MetaTypeInformation> metatypeInformation =
+        (metatypeService == null)
+            ? Collections.emptySet()
+            : Arrays.stream(getContext().getBundles())
+                .filter(b -> bundleLocations.contains(b.getLocation()))
+                .map(metatypeService::getMetaTypeInformation)
+                .collect(Collectors.toSet());
+
+    return services
+        .stream()
+        .filter(
+            service ->
+                hasBundleLocation(service, bundleLocations)
+                    || hasMetatypesForService(service, metatypeInformation))
+        .collect(Collectors.toList());
+  }
 
-    return returnValues;
+  /**
+   * Checks to see if a given service configuration has a bundle location in the set provided.
+   *
+   * @param service the service with a location to check.
+   * @param bundleLocations a set of bundle locations to check.
+   * @return true if the service's location is in bundleLocations, false otherwise.
+   */
+  private boolean hasBundleLocation(Map<String, Object> service, Set<String> bundleLocations) {
+    return Stream.of(service)
+        .filter(m -> m.containsKey("configurations"))
```

Comment:
✏️  Replace these the `filter()` and `map()` with `.map(m -> m.get("configuration")).filter(Objects::nonNull)`. This will avoid double hashing/searching in the map.

Same below for "bundle_location"